### PR TITLE
WIP: Fixed LupIO-TMR and LupV Board 

### PR DESCRIPTION
**Pull request by lhinman19**
The LupV Board was initially connected to the wrong CPU pins (TMR was using incorrect exception codes). Once the correct pins were connected, we were having issues with the LupIO-TMR that was due to some incorrect timing conversions for scheduling events. These issues are now fixed as we are using the correct exception codes and the scheduling of the timer events is occurring properly.

### CODE REVIEW

File: src/python/gem5/components/boards/lupv_board.py [hunk 1]
```diff
@@ -89,36 +92,66 @@ def __init__(
             raise EnvironmentError("RiscvBoard is not compatible with Ruby")
         self.workload = RiscvLinux()
 
+        excep_code = { 'INT_SOFT_SUPER': 1, 'INT_TIMER_SUPER': 5,
+                        'INT_TIMER_MACHINE': 7, 'INT_EXT_SUPER': 9 }
```
Comment (joel-porquet) [inline src/python/gem5/components/boards/lupv_board.py:96]: The indentation every time you have a statement on multiple lines is off in this file.

File: src/python/gem5/components/boards/lupv_board.py [hunk 2]
```diff
@@ -89,36 +92,66 @@ def __init__(
...
+
         # CLINT
         self.clint = Clint(pio_addr=0x2000000)
 
         # PLIC
         self.pic = Plic(pio_addr=0xc000000)
 
+        # LUPIO PIC
+        self.lupio_pic = LupioPIC(pio_addr=0x20002000,
+                                    int_type = excep_code['INT_EXT_SUPER'])
```
Comment (joel-porquet) [inline src/python/gem5/components/boards/lupv_board.py:106]: Here too, and everywhere below. Make sure you don't mix tabs and spaces.

**Review by joel-porquet — COMMENTED** [review]
The PR message is a little confusing. It's not just a fix to the TMR is it? The PR actually deals with the introduction of the TMR+PIC.

**Comment by joel-porquet** [comment]
Oh, is it because you haven't rebased on `lupio/main` first? Could you do that and resubmit the PR so that we see exactly the fixes to TMR and LupV? You wouldn't be able to merge anyway bc of the merge conflicts.

File: src/python/gem5/components/boards/lupv_board.py [hunk 3]
```diff
@@ -269,6 +270,10 @@ def generate_device_tree(self, outdir: str) -> None:
         Creates two files in the outdir: 'device.dtb' and 'device.dts'
         :param outdir: Directory to output the files
         """
+        excep_code = { 'INT_SOFT_SUPER': 1, 'INT_TIMER_SUPER': 5,
+                       'INT_TIMER_MACHINE': 7, 'INT_EXT_SUPER': 9,
+                       'INT_EXT_MACHINE': 11 }
+
```
Comment (joel-porquet) [inline src/python/gem5/components/boards/lupv_board.py:276]: Would you be able to declare this dictionary in such a way that both functions have access to it?

File: src/python/gem5/components/boards/lupv_board.py [hunk 4]
```diff
@@ -101,35 +102,35 @@ def __init__(
         self.pic = Plic(pio_addr=0xc000000)
 
         # LUPIO PIC
-        # The interrupt type is INT_EXT_SUPER
         self.lupio_pic = LupioPIC(pio_addr=0x20002000,
-                                    int_type = excep_code['INT_EXT_SUPER'])
+                                  int_type = excep_code['INT_EXT_SUPER'])
```
Comment (joel-porquet) [inline src/python/gem5/components/boards/lupv_board.py:106]: Ideally, the fixing of the indentation should be a separate commit.

**Review by joel-porquet — COMMENTED** [review]
- Both commit messages are misleading. The first commit is not about adding a new LupIO-TMR device, it's about modifying it. The second commit is not about introducing TMR+PIC in the platform since they're already there, it seems to be about improving the platform's code.
- Ideally, your commits should be split into multiple commits. In the first one about the TMR, you have multiple different logical changes: 1. making the type of interrupt generic, 2. not having a start cycle, 3. Fixing the indentation in the switch/case, 4. Fixing the logic to stop a currently running timer.

Look into `git add --interactive` to separate modifications to the same file into multiple commits. I can do a demo for you if that helps.

File: src/python/gem5/components/boards/lupv_board.py (hunk 3 above)
Comment (powerjg) [reply src/python/gem5/components/boards/lupv_board.py:276]: agreed. I'd put it as a class member

File: src/dev/lupio/lupio_tmr.cc [hunk 5]
```diff
@@ -51,15 +51,9 @@ LupioTMR::LupioTMR(const Params &params) :
     nThread(params.num_threads),
     tmrEvent([this]{ lupioTMRCallback(); }, name()),
     startCycle(0)
-{
-    start = curTick();
```
Comment (joel-porquet) [inline src/dev/lupio/lupio_tmr.cc:55]: If variable `start` becomes unused entirely, you'd probably want to remove it from the class definition in the header file.

File: src/dev/lupio/lupio_tmr.hh [hunk 6]
```diff
@@ -54,7 +54,7 @@ class LupioTMR : public BasicPioDevice
 
     Tick start = 0; 
     Tick next = 0;
```
Comment (joel-porquet) [inline src/dev/lupio/lupio_tmr.hh:57]: What are these two variables for?

**Review by joel-porquet — COMMENTED** [review]
Apart from a couple variables that probably need to be removed for the TMR class, the request looks pretty good :)

**Comment by joel-porquet** [comment]
Just added the missing commit and tested with a non-SMP kernel. Will merge to `lupio/main` so that I can debug the SMP branch.
